File: clojure/rotational-cipher/src/rotational_cipher.clj
(ns rotational-cipher
  (:require [clojure.string :as string]))

(defn rotmap [rot alpha]
  (->> alpha
       (cycle)
       (drop rot)
       (zipmap alpha)))

(defn combine-rotmaps [rot & alphas]
  (apply merge (map (partial rotmap rot) alphas)))

(def alphabet "abcdefghijklmnopqrstuvwxyz")

(defn rotate [s r] 
  (let [mapping 
        (combine-rotmaps r alphabet (string/upper-case alphabet))]
    (apply str (map #(mapping % %) s))))

File: haskell/rotational-cipher/src/RotationalCipher.hs
module RotationalCipher (rotate) where
import Data.Char(isLower, isUpper, toUpper)
import Data.Maybe(fromMaybe)

rotate :: Int -> String -> String
rotate n = map helper
  where
  alphabet = "abcdefghijklmnopqrstuvwxyz"
  createMapping a = zip a (drop n . cycle $ a)
  mapping = createMapping alphabet ++ createMapping (map toUpper alphabet)
  helper c = fromMaybe c $ lookup c mapping
